include/tsar/Analysis/Memory/MemorySet.h

```diff
@@ -158,148 +158,78 @@ class MemorySet {
   /// Return iterator that points to the ending of locations.
   const_iterator end() const { return const_iterator(mLocations.end(), 0); }
 
-  /// Union of returned locations contains a specified location.
+  /// Return true if union of set locations contains a specified location.
+  /// These locations will be stored to `Locs`. 
   template<class Ty>
-  llvm::iterator_range<iterator> findContaining(const Ty &Loc) {
+  bool findContaining(const Ty &Loc,
+                      llvm::SmallVectorImpl<LocationTy> &Locs) const {
     auto I = mLocations.find(MemoryInfo::getPtr(Loc));
     if (I == mLocations.end())
-      return make_range(iterator(mLocations.end(), 0),
-                        iterator(mLocations.end(), 0));
-    auto Tail(Loc);
-    bool KnownStartIdx = false;
-    for (std::size_t Idx = 0, StartIdx = 0, EIdx = I->second.size();
-         Idx < EIdx; ++Idx) {
-      auto &Curr = I->second[Idx];
-      if (MemoryInfo::sizecmp(
-              MemoryInfo::getUpperBound(Curr),
-              MemoryInfo::getLowerBound(Tail)) <= 0)
-        continue;
-      if (!KnownStartIdx)
-        StartIdx = Idx;
-      if (MemoryInfo::sizecmp(
-              MemoryInfo::getLowerBound(Curr),
-              MemoryInfo::getLowerBound(Tail)) > 0)
-        return make_range(iterator(mLocations.end(), 0),
-                          iterator(mLocations.end(), 0));
-      MemoryInfo::setLowerBound(MemoryInfo::getUpperBound(Curr) + 1, Tail);
-      if (MemoryInfo::sizecmp(
-              MemoryInfo::getUpperBound(Tail),
-              MemoryInfo::getLowerBound(Tail)) < 0)
-        return make_range(iterator(I, StartIdx), iterator(I, Idx));
+      return false;
+    llvm::SmallVector<Ty, 4> UnionLocs, Tails { Loc };
+    for (auto &Curr : I->second) {
+      if (Tails.empty())
+        break;
+      llvm::SmallVector<Ty, 4> NewTails;
+      for (auto &Tail : Tails) {
+        auto IntOpt = MemoryInfo::intersect(Curr, Tail, nullptr, &NewTails);
+        if (IntOpt.hasValue()) {
+          if (MemoryInfo::getPtr(IntOpt.getValue()))
+            UnionLocs.push_back(Curr);
+          else
+            return false;
+        } else
+          NewTails.push_back(Tail);

```

Comment:
```suggestion
        if (IntOpt && MemoryInfo::getPtr(*IntOpt))
          Locs.push_back(Curr);
        else          
          NewTails.push_back(Tail);
```

Мне кажется, что если пересечение с каким-то участком памяти есть, но построить его не удалось, это не означает, что не будет какого-то другого участка памяти, с которым удастся построить пересечение.

Reply:
С этим согласна. Только не нужно ли все-таки добавлять `Curr` в `UnionLocs`? Идея была в том, что только если объединение каких-то участков содержит `Loc` целиком, то эти участки добавятся в `Locs`. Пока идет процесс проверки, участки временно помещаются в `UnionLocs`.

Reply:
В целом можно и так и так, я предложил это убрать с целью уменьшить количество копирований. Логика такая, что если метод вернул `false`, то смотреть результат не нужно. Для безопасности, можно так не делать, чтобы гарантировать, что изменений не будет. Тогда на ваше усмотрение.